dispatcher/types.py

```diff
@@ -0,0 +1,16 @@
+# SPDX-License-Identifier: MIT
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any, Callable, Coroutine, Dict, List, Protocol, Tuple, TypeVar
+
+if TYPE_CHECKING:
+    T = TypeVar("T")
+    Coro = Coroutine[Any, Any, T]
+    CoroFunc = Callable[..., Coro[Any]]
+
+    BotT = TypeVar("BotT", bound="SupportsListeners")
+
+    class SupportsListeners(Protocol):
+        extra_events: Dict[str, List[CoroFunc]]
+        dispatch: Callable[[Any, str, Tuple[Any], Dict[str, Any]], None]
```

Comment:
I'm undecided wheter to use this definition or this:
```py
class SupportsListeners(Protocol):
    extra_events: Dict[str, List[CoroFunc]]

    def dispatch(self, event: str, *args: Any, **kwargs: Any) -> None:
        ...
```
btw does this work for `Client` and `AutoShardedClient`? Both have a `dispatch` method but not an `extra_events` attribute

Reply:
If your variant works, then i'd prefer it, it's a lot more readable. And no, it doesn't work with usual (unmodified) Client because every property and definition present on the protocol must be also present on the class you're gonna use. Since Clients dont have extra_events prop, they cant be used.

Reply:
Well the workaround for this would be to define another protocol for `Client` (and AutoShardedClient since its is subclass) and to use an `Union` of protocol. (idk if this will work, it should) How does it look to you?

Reply:
Clients dont support listeners, and as such cannot be supported physically. If one subclasses Client and adds listener support to their custom class then type hinting will work automatically without any additional work.

Reply:
The fact that it does not support listeners does not preclude the dispatch of custom events, furthermore users may be able to define events callback using `@Client.event` or directly subclassing `Client` without the needs of a listeners system (ofc it will allow only 1 callback per event but it will still works)

Reply:
something like this works fine
```py
from typing import Protocol, TypeVar, Generic, Union, Coroutine, Any, Callable, Dict, List, Tuple, TYPE_CHECKING

if TYPE_CHECKING:
    import asyncio

    T = TypeVar("T")
    Coro = Coroutine[Any, Any, T]
    CoroFunc = Callable[..., Coro[Any]]

    class BotSupportsDispatch(Protocol):
        extra_events: Dict[str, List[CoroFunc]]

        def dispatch(self, event_name: str, *args: Any, **kwargs: Any) -> None:
            ...

    class ClientSupportsDispatch(Protocol):
        _listeners: Dict[str, List[Tuple[asyncio.Future, Callable[..., bool]]]]

        def dispatch(self, event: str, *args: Any, **kwargs: Any) -> None:
            ...

    BotT = TypeVar("BotT", bound=Union[BotSupportsDispatch, ClientSupportsDispatch])

    class Dispatcher(Generic[BotT]): pass


    import disnake
    from disnake.ext import commands

    x = Dispatcher[commands.Bot]
    y = Dispatcher[disnake.Client]
```

Reply:
I looked at disnake's source and now have a ton of questions to danny (who originally implemented that stuff). Listeners are basically implemented on Clients as `_listeners` property, but can only be created when using `wait_for()`, are temporary, and running custom logic with them looks ugly. The `@event` decorator instead allows you to define arbitrary functions as listeners, but only one. Why? Because some ass thought that setting listeners as properties on Client is clever. Even more shit adds to this system when you're using Bots - now you have three ways of defining listeners, one awful, second extremely limited, and third exclusive to Bots only. "Awesome", i'd say.
Due to all this i think we should not mess up with supporting Client for now. I'll open an issue on disnake's repo soon, and after this is fixed we can return to supporting Clients.

Reply:
FWI i've already opened an issue on disnake (DisnakeDev/disnake#975) and yes i have a lot of questions too https://canary.discord.com/channels/808030843078836254/913779868985090089/1081975518544535612

Reply:
Alr i think this is resolved as of 1fc1280f78e81020c3236c924b0125bc8f94e237